lib/racecar/concurrent_runner.rb

```diff
@@ -0,0 +1,127 @@
+      @config = config
+      @logger = logger
+      @instrumenter = instrumenter
+    end
+
+    def run
+      logger.info "=> Running with max concurrency of #{config.max_concurrency}"
+
+      self.workers = config.max_concurrency.times.each_with_object({}) do |_, workers|
+        pid, reader = run_worker
+
+        logger.info "=> Forked new Racecar consumer with process id #{pid}"
+
+        workers[pid] = reader
+      end
+
+      install_signal_handlers
+
+      wait_for_exit(parent_readers)
+    end
+
+    def stop
+      worker_pids.each do |pid|
+        $stderr.puts "Waiting for child with pid #{pid}"
+        Process.waitpid(pid)
+      end
+
+      $stderr.puts "All children shutdown, exiting"
+
+      raise exception if exception
+    end
+
+    private
+
+    attr_accessor :exception, :workers
+    attr_reader :processor, :config, :logger, :instrumenter
+
+    def install_signal_handlers
+      # Print the consumer config to STDERR on USR1.
+      trap("USR1") { $stderr.puts config.inspect }
+
+      SHUTDOWN_SIGNALS.each do |signal|
+        trap(signal) do
+          $stderr.puts "Received signal in parent: #{signal}, stopping"
+          terminate_child_processes
+        end
+      end
+    end
+
+    def run_worker
+      parent_reader, child_writer = IO.pipe(PIPE_ENCODING, PIPE_ENCODING, binmode: true)
+
+      pid = fork do
+        begin
+          parent_reader.close
+
+          Racecar::Runner.new(processor, config: config, logger: logger, instrumenter: instrumenter).run
+        rescue Exception => e
+          # Allow the parent process to re-raise the exception after shutdown
+          child_writer.set_encoding('binary')
+          child_writer.write(Marshal.dump(e))
+        ensure
+          child_writer.close
+        end
+      end
+
+      child_writer.close
+
+      [pid, parent_reader]
+    end
+
+    def terminate_child_processes
+      $stderr.puts "=> Terminating child processes"
+
+      Process.kill("TERM", *worker_pids)
+    end
+
+    def wait_for_exit(remaining_readers)
+      if remaining_readers.empty?
+        stop
+      else
+        # The call to IO.select blocks until one of our readers is ready for reading, which
+        # could be for one of two reasons:
+        #
+        # - An exception is raised in the child process, in which case we should initiate
+        #   a shutdown;
+        #
+        # - A graceful shutdown was already initiated, and the pipe writer has been closed, in
+        #   which case there is nothing more to do.
+        #
+        ready_readers = IO.select(remaining_readers).first
+
+        first_read = ready_readers.first.read
+        exception_found = !first_read.empty?
+
+        if exception_found
+          self.exception = Marshal.load(first_read)
+          terminate_child_processes
+        end
+
+        ready_readers.each(&:close)
+
+        # Recursively wait for the remaining readers
+        wait_for_exit(remaining_readers - ready_readers)
```

Comment:
This risks blowing the stack if we go too deep. Probably iterate instead, e.g. `until remaining_readers.empty? ... end` and update a variable with the remaining readers.

Reply:
Indeed, this wasn't needed in the end 👍